app/models/public_body.rb

```diff
@@ -855,6 +855,32 @@ def self.with_query(query, tag)
     end
   end
 
+  # This method updates the count columns of the PublicBody that
+  # store the number of "not held", "to some extent successful" and
+  # "both visible and classified" requests.
+  def update_counter_cache
+    success_states = %w(successful partially_successful)
+    basic_params = { public_body_id: id }
+
+    mappings = {
+      info_requests_not_held_count: { awaiting_description: false,
+                                      described_state: 'not_held' },
+      info_requests_successful_count: { awaiting_description: false,
+                                        described_state: success_states },
+      info_requests_visible_classified_count: { awaiting_description: false},
+      info_requests_visible_count: {}
+    }
+
+    updated_counts = mappings.reduce({}) do |memo, mapping|
+      column = mapping.first
+      params = basic_params.clone.update(mapping.last)
```

Comment:
Not a fan of this, and as `.where` are chainable this could be refactored as:

```ruby
info_request_scope = InfoRequest.where(public_body_id: id).is_searchable
...
updated_counts = mappings.reduce({}) do |memo, (column, params)|
  memo[column] = info_request_scope.where(params).count 
```

Note the block arguments is expanding the hash to remove `.first`/`.last` calls too.

Reply:
Made this refactoring in d3bd013ee.